third_party/amd/lib/TritonAMDGPUTransforms/StreamPipeline.cpp

```diff
@@ -196,54 +209,79 @@ class StreamPipeliner {
-    //   local_load:   stage=i
-    //   compute:      stage=i
-    //   tail:         stage=i
-    config[SCHED_LOCAL_STORE] = {lastStage - 1, schedule.clusters.newAtBack()};
-    auto cluster1 = schedule.clusters.newAtBack();
-    config[SCHED_GLOBAL_LOAD] = {0, cluster1};
-    config[SCHED_LOCAL_LOAD] = {lastStage, cluster1};
-    config[SCHED_COMPUTE] = {lastStage, cluster1};
-    config[SCHED_TAIL] = {lastStage, schedule.clusters.newAtBack()};
+LogicalResult StreamPipeliner::initSchedule(int maxIndirectionLevel) {
+
+  bool pairedGlobalLoadLocalStore = stages[SCHED_LOCAL_STORE] == 0;
+  if (stages[SCHED_LOCAL_STORE] < 0) {
+    stages[SCHED_LOCAL_STORE] = numStages + stages[SCHED_LOCAL_STORE];
   } else {
-    // Streaming Schema cluster order and staging for single-buffer.
-    // for i in (...):
-    //   global_loads: stage=i+1
-    //   local_load:   stage=i
-    //   compute:      stage=i
-    //   local_stores: stage=i+1
-    //   tail:         stage=i
-    auto cluster0 = schedule.clusters.newAtBack();
-    config[SCHED_GLOBAL_LOAD] = {0, cluster0};
-    config[SCHED_LOCAL_LOAD] = {lastStage, schedule.clusters.newAtBack()};
-    config[SCHED_COMPUTE] = {lastStage, cluster0};
-    config[SCHED_LOCAL_STORE] = {lastStage - 1, schedule.clusters.newAtBack()};
-    config[SCHED_TAIL] = {lastStage, schedule.clusters.newAtBack()};
+    stages[SCHED_LOCAL_STORE] += maxIndirectionLevel;
+  }
+
+  if (stages[SCHED_LOCAL_LOAD] < 0)
+    stages[SCHED_LOCAL_LOAD] = numStages + stages[SCHED_LOCAL_LOAD];
+
+  if (stages[SCHED_GLOBAL_LOAD] > stages[SCHED_LOCAL_STORE] ||
+      stages[SCHED_LOCAL_STORE] > stages[SCHED_LOCAL_LOAD]) {
+    LDBG("Invalid schedule: GLOBAL_LOAD stage = "
+         << stages[SCHED_GLOBAL_LOAD]
+         << ", LOCAL_STORE stage = " << stages[SCHED_LOCAL_STORE]
+         << ", LOCAL_LOAD stage = " << stages[SCHED_LOCAL_LOAD]
+         << ", COMPUTE stage = " << stages[SCHED_COMPUTE]);
+    return failure();
+  }
+
+  // Calculate the number of buffers needed for each load.
+  // TODO: Use the precise number of buffers needed by the particular load.
+  numBuffers =
+      std::max(1, stages[SCHED_LOCAL_LOAD] - stages[SCHED_LOCAL_STORE]);
+  LDBG("deduced max shared memory buffer number = " << numBuffers);
+
+  // If tt.load and ttg.local_store are in the same stage
+  //   spread them apart to allow overlap with compute
+  // else
+  //   Initiate ttg.local_store before tt.load
+  int globalLoadCluster = 0;
+  int localStoreCluster = 2;
+  if (!pairedGlobalLoadLocalStore) {
+    globalLoadCluster = 2;
+    localStoreCluster = 1;
+  }
+
+  // If ttg.local_load and ttg.local_store are in the same stage
+  //   spread them apart to allow overlap with compute
+  // else if they share the buffer
+  //   ttg.local_load must come first
+  // else
+  //   schedule ttg.local_load in the middle
+  int localLoadCluster = globalLoadCluster;
+  if (stages[SCHED_LOCAL_LOAD] == stages[SCHED_LOCAL_STORE]) {
+    localLoadCluster = 2;
```

Comment:
In the case of 
stage0: GR, LW, LR
stage1: DOT

LW and LR are assigned to cluster 2. This leads to the following IR
```mlir
load A
load B
dot
local_store A
local_load A
local_store B
local_load B
```
Then in this case, A and B (assuming they have the same shape, thus same size in LDS) should reuse the same LDS buffer. And line 236 also indicates there is 1 buffer in LDS. But it uses double buffers in LDS. I'm wondering
- If this is expected?
- If not, is there something wrong with cluster assignment? or Shared-allocation pass?

Reply:
In the case of 
stage0: GR
stage1: LW, LR, DOT

LW is in cluster 1, GR, LR, and DOT are all in cluster 2. 
I can see you want LW to proceed GR so that they can share the same register buffer. However, GR and LR should not be in the same cluster, since we want to issue GR as early as possible.

Reply:
It depends.. sometimes it may be more performant in this order. 

I think this is a case where micro-scheduling will adjust for better placement.

But the first scenario you mention does seem less optimal. I'll adjust that.